### PR TITLE
feat: Implements HTML as a column type in table widget.

### LINKED ISSUE #13681: [Feature]: Support HTML as a column type in the table widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Allow using html as a column type in the table widget

### Why should this be worked on?

Enhances our table widget to render custom html making the tables look better

Front logo [Front conversations](https://app.frontapp.com/open/top_b0vnw)

### PR DESCRIPTION
**Pull request by rahulbarwal**
## Description

Adds support for rendering HTML content in Table Widget V2 columns with proper sanitization and filtering capabilities.

- Adds new HTML column type with support for:
  - Rendering sanitized HTML content using Interweave
  - Search/filter functionality that ignores HTML tags
  - Sort functionality based on text content
  - CSV export with proper HTML content handling
  - Added unit tests covering HTML column functionality

Fixes #13681
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.Table"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12270923512>
> Commit: 5f873fca08bc762707190c183c8cfd0ec403b978
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12270923512&attempt=2" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Table`
> Spec:
> <hr>Wed, 11 Dec 2024 07:23:01 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Release Notes

- **New Features**
  - Introduced support for HTML content in table cells with the new `HTMLCell` component.
  - Enhanced filtering and sorting functionalities to handle HTML columns effectively.
  - Added a feature flag to enable or disable HTML column types.
  - Implemented new configurations for column types, including HTML options.
  - Added analytics tracking for HTML cell usage.

- **Bug Fixes**
  - Improved HTML sanitization to prevent unsafe content rendering.

- **Tests**
  - Expanded test coverage for HTML handling in table widgets, including rendering, filtering, and sorting scenarios.
  - Added comprehensive testing for the `HTMLCell` component and its interaction with HTML content.
  - Introduced Cypress end-to-end tests for HTML cell functionality within the table widget.

### CODE REVIEW

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    isHidden,
+    textColor,
+    textSize,
+    value,
+    verticalAlignment,
+  } = props;
+
+  const interweaveCompatibleValue = useMemo(convertToInterweaveString(value), [
+    value,
+  ]);
```
[the remark this answers is not in the document]
Comment (rahulbarwal) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell.tsx:94]: Done.

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/utils.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  (tags: string[]) => {
+    AnalyticsUtil.logEvent("TABLE_WIDGET_V2_HTML_CELL_USAGE", {
+      tags: tags,
+    });
+  },
+  1000,
+  { leading: true, trailing: false, maxWait: 5000 },
+);
+
+// Function to extract unique HTML tags from a string
```
Comment (jsartisan) [inline app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/utils.tsx:15]: I used to do same earlier 😅 but such comments are useless. It's already obvious you are extracting HTML tags from a string by looking at the definition of the function.

File: app/client/src/widgets/TableWidgetV2/component/header/actions/Utilities.ts [hunk 3]
```diff
@@ -35,7 +36,11 @@ export const transformTableDataIntoCsv = (props: {
             ? value.replace("\n", " ")
             : value;
 
-        if (isString(value) && value.includes(",")) {
+        const shouldQuote =
```
Comment (jsartisan) [inline app/client/src/widgets/TableWidgetV2/component/header/actions/Utilities.ts:39]: What does "shouldQuote" means, it's hard to tell why here.

File: app/client/src/widgets/TableWidgetV2/widget/derived.js [hunk 4]
```diff
@@ -781,32 +800,51 @@ export default {
...
+          displayText = column.displayText;
+        }
+
+        acc[column.alias] = displayText;
+
+        return acc;
+      }, {});
+
+      /*
+       * We don't want html tags and inline styles to match in search
```
Comment (jsartisan) [inline app/client/src/widgets/TableWidgetV2/widget/derived.js:818]: Now this is a good comment, you are telling "why" you are doing this.

File: app/client/src/widgets/TableWidgetV2/widget/derived.js [hunk 5]
```diff
@@ -781,32 +800,51 @@ export default {
...
+        ...row,
+        ...labelValuesForSelectCell,
+        ...displayTextValues,
+        ...htmlValues,
       };
+      const htmlColumns = columnsWithHTML.map((column) => column.alias);
 
       if (searchKey) {
-        isSearchKeyFound = [
+        const arr = [
```
Comment (jsartisan) [inline app/client/src/widgets/TableWidgetV2/widget/derived.js:838]: const arr is not a good name. You can use `combinedRowContent or just do:

```js
isSearchKeyFound = [
  ...Object.values(_.omit(displayedRow, hiddenColumns)),
  ...Object.values(
    _.omit(originalRow, [...hiddenColumns, ...htmlColumns]),
  ),
].includes(searchKey)
```

File: app/client/src/widgets/TableWidgetV2/widget/index.tsx [hunk 6]
```diff
@@ -911,6 +914,38 @@ class TableWidgetV2 extends BaseWidget<TableWidgetProps, WidgetState> {
       //dont neet to batch this since single action
       this.hydrateStickyColumns();
     }
+
+    /**
+     * Why we are doing this?
```
Comment (jsartisan) [inline app/client/src/widgets/TableWidgetV2/widget/index.tsx:919]: 💙

File: app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/Data/ColumnType.ts [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    value: ColumnTypes.URL,
+  },
+  {
+    label: "Video",
+    value: ColumnTypes.VIDEO,
+  },
+];
+
+const columnTypeWithHtml = [
+  ...ColumnTypeOptions.slice(0, 4),
```
Comment (jsartisan) [inline app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/Data/ColumnType.ts:72]: Why 4? or not 5 or 6? What is significance of 4 number? Maybe we can use a constant whose name tell the context a little better?

**Review by jsartisan — COMMENTED** [review]
Overall looks ok. but left few comments that can improve the readability of code.

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/index.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+import Interweave from "interweave";
+import { isEqual } from "lodash";
+import React, { useEffect, useMemo, useRef } from "react";
+import styled from "styled-components";
+import LinkFilter from "widgets/TextWidget/component/filters/LinkFilter";
+import type { BaseCellComponentProps } from "../../Constants";
+import { CellWrapper } from "../../TableStyledWrappers";
+import { extractHTMLTags, getRenderMode, sendHTMLCellAnalytics } from "./utils";
+
+const HTMLContainer = styled.div`
```
[the remark this answers is not in the document]
Comment (jsartisan) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/index.tsx:11]: Btw, I am saying not to do this in this PR, this is just an FYI

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/index.tsx [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+  const previousTagsRef = useRef<string[]>([]);
+
+  const interweaveCompatibleValue = useMemo(() => {
+    if (value === null || value === undefined) return "";
+
+    return String(value);
+  }, [value]);
+
+  const extractedTags = useMemo(() => {
```
[the remark this answers is not in the document]
Comment (rahulbarwal) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/index.tsx:105]: Done.

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/index.tsx [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    if (!interweaveCompatibleValue) return [];
+
+    return extractHTMLTags(interweaveCompatibleValue);
+  }, [interweaveCompatibleValue]);
+
+  useEffect(() => {
+    const areTagsChanged = !isEqual(
+      extractedTags.sort(),
+      previousTagsRef.current.sort(),
+    );
```
[the remark this answers is not in the document]
Comment (rahulbarwal) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/index.tsx:115]: Thanks! added changes.

File: app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/index.tsx [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    const isRenderModeValid = ["DEPLOYED", "EDITOR"].includes(
```
[the remark this answers is not in the document]
Comment (rahulbarwal) [reply app/client/src/widgets/TableWidgetV2/component/cellComponents/HTMLCell/index.tsx:116]: Yes, makes sense. My bad, removing this piece of code.

File: app/client/src/widgets/TableWidgetV2/component/header/actions/Utilities.ts (hunk 3 above)
Comment (rahulbarwal) [reply app/client/src/widgets/TableWidgetV2/component/header/actions/Utilities.ts:39]: HTML values are interpreted as multiline strings. If they aren’t enclosed in backticks, the output can become disorganized and visually unappealing.
Adding a comment as well.

File: app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/Data/ColumnType.ts (hunk 7 above)
Comment (rahulbarwal) [reply app/client/src/widgets/TableWidgetV2/widget/propertyConfig/PanelConfig/Data/ColumnType.ts:72]: Alphabetically sorted! will add a comment.

File: app/client/src/widgets/TableWidgetV2/widget/utilities.ts [hunk 12]
```diff
@@ -771,6 +771,14 @@ export const getColumnType = (
...
       return ColumnTypes.CHECKBOX;
     case "string":
+      // Check for HTML first
+      const isHTML = /<[^>]*>/.test(columnValue as string);
+
+      if (isHTML) {
+        return ColumnTypes.HTML;
+      }
+
+      // Then check for date
```
[the remark this answers is not in the document]
Comment (rahulbarwal) [reply app/client/src/widgets/TableWidgetV2/widget/utilities.ts:781]: Done.

File: app/client/src/widgets/TableWidgetV2/widget/utilities.ts [hunk 13]
```diff
@@ -770,12 +770,23 @@ export const getColumnType = (
       return ColumnTypes.NUMBER;
     case "boolean":
       return ColumnTypes.CHECKBOX;
-    case "string":
-      return dateFormatOptions.some(({ value: format }) =>
+    case "string": {
+      const isHTML = /<[^>]*>/.test(columnValue as string);
```
Comment (jacquesikot) [inline app/client/src/widgets/TableWidgetV2/widget/utilities.ts:774]: This method doesn't seem exhaustive enough, there's a high chance of false positives with this, can we use something like this?

function detectHTML(input: string): boolean {
  const div = document.createElement("div");
  div.innerHTML = input;
  return div.innerHTML !== input; // If innerHTML modifies input, it's likely HTML
}

Comment (jacquesikot) [reply app/client/src/widgets/TableWidgetV2/widget/utilities.ts:774]: Or a stricter regex?

Comment (rahulbarwal) [reply app/client/src/widgets/TableWidgetV2/widget/utilities.ts:774]: We analyzed this, and the current check is sufficient to validate html. 
1. `Testing <span> text </text>`
2. `<strong> Rahul `

Invalid HTML, text not starting with a tag are all rendered by browser correctly and will not pass the aforementioned validity function. 
It's best to stay put with current implementation. Let me know if you disagree.

File: app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/columnTypes/HTMLCell_spec.ts [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+      });
+      table.SortColumn("status", "desc");
+      table.ReadTableRowColumnData(0, 3, "v2").then(($cellData) => {
+        expect($cellData).to.include("Suspended");
+      });
+    });
+  },
+);
+
+const htmlTableData = [
```
Comment (sagar-qa007) [inline app/client/cypress/e2e/Regression/ClientSide/Widgets/TableV2/columnTypes/HTMLCell_spec.ts:81]: Can we please put this data in fixture folder as file?

**Review by sagar-qa007 — CHANGES_REQUESTED** [review]

**Review by jsartisan — DISMISSED** [review]

**Review by sagar-qa007 — APPROVED** [review]

**Pull request merged**